### PR TITLE
FOCS support for part based fleet upkeep

### PR DESCRIPTION
**Pull request by dbenage-cx**
~~PR is intended for testing of fleet upkeep utilizing number of currently owned parts.
Initial change from 1% per ship to 0.2% per part.
While there are various threads discussing changes to upkeep, the most recent revolves around [Influence & Happiness](http://www.freeorion.org/forum/viewtopic.php?f=5&t=9998)~~

Changed to only add support for such a change.

### CODE REVIEW

File: UI/MapWnd.cpp [hunk 1]
```diff
@@ -5779,16 +5779,36 @@ void MapWnd::RefreshFleetResourceIndicator() {
     const std::set<int>& this_client_known_destroyed_objects = GetUniverse().EmpireKnownDestroyedObjectIDs(empire_id);
 
     int total_fleet_count = 0;
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp:5782]: if adding other counts of ships (not actually counting fleets), this should be renamed to total_ship_count

File: UI/MapWnd.cpp [hunk 2]
```diff
@@ -5779,16 +5779,36 @@ void MapWnd::RefreshFleetResourceIndicator() {
...
+            total_ship_slot_count += hull_type->NumSlots();
+        }
     }
 
     m_fleet->SetValue(total_fleet_count);
     m_fleet->ClearBrowseInfoWnd();
     m_fleet->SetBrowseModeTime(GetOptionsDB().Get<int>("UI.tooltip-delay"));
     m_fleet->SetBrowseInfoWnd(boost::shared_ptr<GG::BrowseInfoWnd>(
-        new TextBrowseWnd(UserString("MAP_FLEET_TITLE"), UserString("MAP_FLEET_TEXT"))));
+        new TextBrowseWnd(UserString("MAP_FLEET_TITLE"),
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp]: Can you make this a separate pull request?

File: parse/IntComplexValueRefParser.cpp [hunk 3]
```diff
@@ -102,6 +102,13 @@ namespace parse {
                     ) [ _val = new_<ValueRef::ComplexVariable<int> >(_a, _b, _c, _f, _d, _e) ]
                 ;
 
+            ship_parts_owned
```
Comment (geoffthemedio) [inline parse/IntComplexValueRefParser.cpp:105]: not necessary now, but planning to add a parameter to specify a part type would be good

File: Empire/Empire.cpp [hunk 4]
```diff
@@ -3429,6 +3429,17 @@ int Empire::TotalShipsOwned() const {
     return counter;
 }
 
+int Empire::TotalShipPartsOwned() const {
```
Comment (geoffthemedio) [inline Empire/Empire.cpp:3446]: You can't recalculate this every time the value is needed. It will need to be store in the empire as state info, and updated once per turn in UpdateOwnedObjectCounters

File: UI/MapWnd.cpp (hunk 2 above)
Comment (dbenage-cx) [reply UI/MapWnd.cpp]: Done. PR #1210 (will drop for this PR when updated with other changes requested)

File: UI/MapWnd.cpp (hunk 1 above)
Comment (dbenage-cx) [reply UI/MapWnd.cpp:5782]: Fixed in other PR

**Comment by Vezzra** [comment]
> number of currently owned parts

IIRC @geoffthemedio had already added FOCS keywords that returned the total number of slots and number of used slots on a ship design, what are "owned parts" supposed to be?

**Comment by Vezzra** [comment]
@freeorion/ai-team, what adjustments to the AI are required for this?

**Comment by MatGB** [comment]
The AI gets the upkeep from here [freeorion/AIDependencies.py at master · freeorion/freeorion](https://github.com/freeorion/freeorion/blob/master/default/python/AI/AIDependencies.py#L54) but it's backend coded to think that's a per ship cost. I've started my first game with this now and given it's very close to what I wanted to try out (but could never get working, fairly sure Geoff's FOCS was PartsInShipDesign but it was so long ago I last tried to play around with that) unsurprisingly I like it.

If we do have an upkeep-as-govt-choice mechanic something like this is far preferable to the current blunt edged setup and this appears to work fine on initial pass. Only on turn 70 with 14 ships total though.

File: parse/IntComplexValueRefParser.cpp (hunk 3 above)
Comment (dbenage-cx) [reply parse/IntComplexValueRefParser.cpp:105]: Added

File: Empire/Empire.cpp (hunk 4 above)
Comment (dbenage-cx) [reply Empire/Empire.cpp:3446]: Fixed

**Comment by dbenage-cx** [comment]
Added ShipPartClassOwned to support different upkeep per part class.
e.g. For upkeep to increase only for short range weapon parts
`(1 + 0.004 * ShipPartClassOwned empire = Source.Owner class = ShortRange)`

No work was done towards AI yet, would like to finalize a formula for upkeep beforehand.

> what are "owned parts" supposed to be?

Current number of ship parts for all existing ships, possibly limited to those owned by a specific empire and/or of a specific part type.  In the use case here, it is for the total number of currently owned parts for the source empire.

**Comment by MatGB** [comment]
More feedback: I've thought for a long time but this now confirms it that we need something in the UI displaying the current Upkeep rates, it's confusing enough for new players we should be trying to make it more discoverable, and my brain keeps seizing up when I ask it to work out the fleet upkeep with this system. 

Having it in the Fleet Metrics mouseover is a bare minimum, also having it in the production screen somewhere, perhaps in the box where it gives your total production and prod in this resourcegroup?

Same for colony upkeep, in the same place in production if that's possible, then a colony count per species and a total colony at the bottom of the census pane? (that's outside the remit of this PR, obviously).

File: Empire/Empire.cpp [hunk 5]
```diff
@@ -3386,6 +3388,20 @@ void Empire::UpdateOwnedObjectCounters() {
...
+        // parts
+        const ShipDesign* design = ship->Design();
+        if (!design)
+            continue;
+        const std::vector<std::string>& parts = design->Parts();
+        for (const std::string& part : parts) {
+            const PartType* part_type = GetPartType(part);
+            if (!part_type)
+                continue;
+            if (!part_type->Name().empty())
```
Comment (geoffthemedio) [inline Empire/Empire.cpp]: should check for blank name before looking up, not after, as the latter will never return an empty name if the lookup worked.

File: Empire/Empire.cpp [hunk 6]
```diff
@@ -3372,9 +3372,11 @@ void Empire::ResetMeters() {
 }
 
 void Empire::UpdateOwnedObjectCounters() {
-    // ships of each species and design
+    // ships of each species and design, parts by type and class
     m_species_ships_owned.clear();
     m_ship_designs_owned.clear();
+    m_ship_parts_owned.clear();
```
Comment (geoffthemedio) [inline Empire/Empire.cpp]: Add a comment explaining the different containers.... one indexed by part name, one by part class, summing either could be used for total ship parts count.

File: Empire/Empire.h [hunk 7]
```diff
@@ -567,10 +567,14 @@ class FO_COMMON_API Empire {
     void                        UpdateOwnedObjectCounters();
 
     int                         TotalShipsOwned() const;
+    /** Total number of parts for all owned ships in this empire */
+    int                         TotalShipPartsOwned() const;
     int                         TotalBuildingsOwned() const;
 
     std::map<std::string, int>& SpeciesShipsOwned()     { return m_species_ships_owned; }
```
Comment (geoffthemedio) [inline Empire/Empire.h:573]: make everything line up horizontally...

File: Empire/Empire.cpp [hunk 8]
```diff
@@ -3386,6 +3388,20 @@ void Empire::UpdateOwnedObjectCounters() {
         if (!ship->SpeciesName().empty())
             m_species_ships_owned[ship->SpeciesName()]++;
         m_ship_designs_owned[ship->DesignID()]++;
+        // parts
+        const ShipDesign* design = ship->Design();
+        if (!design)
+            continue;
+        const std::vector<std::string>& parts = design->Parts();
```
Comment (geoffthemedio) [inline Empire/Empire.cpp]: This might be better moved into ShipDesign. Could add a ShipDesign::PartCount function that returns, or takes in by reference and fills, maps from PartClass and from string to int? That info could also be cached in the ShipDesign in a mutable member after it is calculated...

File: Empire/Empire.cpp [hunk 9]
```diff
@@ -3372,9 +3372,11 @@ void Empire::ResetMeters() {
 }
 
 void Empire::UpdateOwnedObjectCounters() {
-    // ships of each species and design
+    // ships of each species and design, parts by type and class
     m_species_ships_owned.clear();
     m_ship_designs_owned.clear();
+    m_ship_parts_owned.clear();
+    m_ship_part_class_owned.clear();
     for (std::map<int, TemporaryPtr<UniverseObject> >::iterator ship_it = Objects().ExistingShipsBegin();
```
Comment (geoffthemedio) [inline Empire/Empire.cpp:3375]: Probably doesn't matter for a once-per-turn calculation, but could calculate number of parts in each design once, and then multiply by the number of each design the empire owns, instead of processing each ship separately.

File: parse/IntComplexValueRefParser.cpp [hunk 10]
```diff
@@ -102,6 +102,30 @@ namespace parse {
...
+                =   (
+                            tok.ShipPartsOwned_ [ _a = construct<std::string>(_1) ]
+                        >-( parse::label(Empire_token) >    int_value_ref [ _b = _1 ] )
+                        >-( parse::label(Name_token)   >    string_value_ref [ _d = _1 ] )
+                    ) [ _val = new_<ValueRef::ComplexVariable<int>>(_a, _b, _c, _f, _d, _e) ]
+                ;
+
+            ship_part_class_owned
+                = (
+                            tok.ShipPartClassOwned_ [ _a = construct<std::string>(_1) ]
```
Comment (geoffthemedio) [inline parse/IntComplexValueRefParser.cpp]: probably don't need a separate name for this value in FOCS... could construct the parts of class owned variant when a class is specified, and the parts by name owned variant when a string is specified. Shouldn't be any ambiguity between these...

File: parse/IntComplexValueRefParser.cpp [hunk 11]
```diff
@@ -102,6 +102,30 @@ namespace parse {
...
+                        >-( parse::label(Empire_token) >    int_value_ref [ _b = _1 ] )
+                        >-( parse::label(Class_token)  >
```
Comment (geoffthemedio) [inline parse/IntComplexValueRefParser.cpp]: enum_parser<ShipPartClass>() already exists

File: universe/ValueRef.cpp [hunk 12]
```diff
@@ -1278,6 +1278,8 @@ namespace {
             return empire->SpeciesShipsProduced();
         if (parsed_map_name == "SpeciesShipsScrapped")
             return empire->SpeciesShipsScrapped();
+        if (parsed_map_name == "ShipPartsOwned" || parsed_map_name == "ShipPartClassOwned")
```
Comment (geoffthemedio) [inline universe/ValueRef.cpp]: should these really both be calling the same function?

File: parse/IntComplexValueRefParser.cpp [hunk 13]
```diff
@@ -102,6 +102,30 @@ namespace parse {
...
+                            ( tok.ShortRange_       | tok.FighterBay_   | tok.FighterWeapon_
+                            | tok.Shield_           | tok.Armour_
+                            | tok.Troops_           | tok.Detection_    | tok.Stealth_
+                            | tok.Fuel_             | tok.Colony_       | tok.Speed_
+                            | tok.General_          | tok.Bombard_      | tok.Research_
+                            | tok.Industry_         | tok.Trade_        | tok.ProductionLocation_
+                            ) [ _d = new_<ValueRef::Constant<std::string> >(_1) ]
```
Comment (geoffthemedio) [inline parse/IntComplexValueRefParser.cpp]: should probably be consistent about whether to use >> or > > when closing nested templates

**Review by geoffthemedio — CHANGES_REQUESTED** [review]

**Comment by Morlic-fo** [comment]
> @freeorion/ai-team, what adjustments to the AI are required for this?

I will try to look into this tomorrow.

File: parse/IntComplexValueRefParser.cpp (hunk 11 above)
Comment (dbenage-cx) [reply parse/IntComplexValueRefParser.cpp]: `as_string` seems to return `std::vector<char>` instead of `std::string`.  I'm not certain how to overcome that in this context, hence the use of int. (edit: I'm not certain what value is expected here, but it seems `as_string` is not suitable by itself)

**Comment by dbenage-cx** [comment]
Rebased commits, removing `ShipPartClassOwned` and adding a `class` value for `ShipPartsOwned`.
If both `name` and `class` are specified, only the first is utilized.

File: parse/IntComplexValueRefParser.cpp (hunk 11 above)
Comment (geoffthemedio) [reply parse/IntComplexValueRefParser.cpp]: hmm, looks like I had the same idea before and gave up on getting it to work, and did what you've done here (for part_class_in_ship_design)

Maybe a solution here, but probably not worth the effort: http://www.doing-stuff.org/2015/04/obtaining-both-the-attribute-and-the-raw-string-in-boostspirit/

**Comment by geoffthemedio** [comment]
@MatGB Related: https://github.com/freeorion/freeorion/issues/1211

**Comment by Morlic-fo** [comment]
What stops me from spamming COMsats or generally empty hulls as cannon fodder?

Such hulls definitely stay relevant even with fighters as they protect the main fleet from capital ship fire in the first turn before fighters are launched.

**Comment by Vezzra** [comment]
> What stops me from spamming COMsats or generally empty hulls as cannon fodder?

Good catch. That means when calculating the cost increase factor we either need to either:

a) still also need to factor in the number of ships (so even ships without any parts on their design get counted)

b) consider the part slots regardless if their are used or not.

**Comment by MatGB** [comment]
c) have a different specific upkeep for the Base hull so it can be used for intended purpose (outpost bases and colony bases), can be used for edge case uses (comsats/chaff/setting system to aggressive) but is too costly to use in large numbers.

I've thought an actual cost per turn in PP for the system it's in would be doable, each base hull in a system reduces PP output in that system by an amount (probably 1).

We could have an individual hull count as a part for counting but it'd still be a very low upkeep amount compared to current, I'm opposed to using slots available instead of slots used because part of my vision for this was to reduce the need players feel to use every slot in every design: if you don't need a fuel tank, don't design it in, but if cost is per slot then you might as well fill everything.

(we do need to work out a better UI for setting a system to aggressive than having a comsat in orbit, but that's outside this remit)

FWIW, I'm now into my 4th game with the original version of this PR that had the upkeep set to 0.2% per part, I *really* like it, it's less upkeep in the early game but a substantial increase in the late game and I'm finding myself using more smaller hulls way into late game in a way I didn't before.

**Comment by Morlic-fo** [comment]
> c) have a different specific upkeep for the Base hull so it can be used for intended purpose (outpost bases and colony bases), can be used for edge case uses (comsats/chaff/setting system to aggressive) but is too costly to use in large numbers.

This is not a solution. I would simply use other cheap hulls. For supporting a battle fleet, I need to use mobile hulls anyway.

**Comment by dbenage-cx** [comment]
Expect we probably want upkeep applied through effects rather than appended to `buildcost` in the definitions.  Such effects would apply to the previous linked empire meters.
